libraries-3/src/main/java/com/baeldung/jnats/NatsClient.java

```diff
@@ -1,123 +1,144 @@
 package com.baeldung.jnats;
 
 import java.io.IOException;
+import java.time.Duration;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
 
+import io.nats.client.*;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import io.nats.client.AsyncSubscription;
-import io.nats.client.Connection;
-import io.nats.client.Message;
-import io.nats.client.Nats;
-import io.nats.client.Options;
-import io.nats.client.Subscription;
-import io.nats.client.SyncSubscription;
-
 public final class NatsClient {
 
     private final String serverURI;
 
-    private final Connection natsConnection;
-
-    private final Map<String, Subscription> subscriptions = new HashMap<>();
+    private final Map<Subscription, Dispatcher> dispatcherBySubscription = new HashMap<>();
 
     private final static Logger log = LoggerFactory.getLogger(NatsClient.class);
 
+    private Connection natsConnection;
+
     NatsClient() {
-        this.serverURI = "jnats://localhost:4222";
-        natsConnection = initConnection(serverURI);
+        this(null);
     }
 
     public NatsClient(String serverURI) {
-        if ((serverURI != null) && (!serverURI.isEmpty())) {
+        if (serverURI == null || serverURI.isEmpty()) {
+            this.serverURI = "nats://localhost:4222";
+        }
+        else {
             this.serverURI = serverURI;
-        } else {
-            this.serverURI = "jnats://localhost:4222";
         }
+        initConnection();
+    }
 
-        natsConnection = initConnection(serverURI);
+    static class CustomErrorListener implements ErrorListener {
+        @Override
+        public void errorOccurred(Connection conn, String error) {
+            log.error("Error Occurred: {}", error);
+        }
+
+        @Override
+        public void exceptionOccurred(Connection conn, Exception exp) {
+            log.error("Exception Occurred: {}", exp.toString());
+        }
+    }
+    private void initConnection() {
+        try {
+            Options options = new Options.Builder()
+                .server(serverURI)
+                .connectionListener((connection, event) -> log.info("Connection Event: {}", event.toString()))
+                .errorListener(new CustomErrorListener())
+                .build();
+            natsConnection = Nats.connect(options);
+        } catch (IOException | InterruptedException e) {
+            log.error("Error connecting to NATs! {}", e.toString());
+        }
```

Comment:
When we catch InterruptedException it should always be followed by Thread.currentThread().interrupt() inside the catch block to restore the interrupt status